horovod/common/elastic.py

```diff
@@ -0,0 +1,166 @@
+
+
+notification_manager = WorkerNotificationManager()
+
+
+class State(object):
+    """State representation used for tracking in memory state across workers.
+
+    Args:
+        bcast_object: Function used to broadcast a variable from rank 0 to the other workers.
+        get_rank: Function that returns the current rank of this worker.
+    """
+    def __init__(self, bcast_object, get_rank):
+        self._bcast_object = bcast_object
+        self._rank = get_rank
+        self._host_messages = queue.Queue()
+        self._last_updated_timestamp = 0
+        self._reset_callbacks = []
+
+    def register_reset_callbacks(self, callbacks):
+        """Register callbacks that will be invoked following a reset event (worker added or removed).
+
+        For example, a common use of a reset callback would be to update the learning rate scale with the
+        new number of workers.
+
+        Args:
+            callbacks: list of functions to execute.
+        """
+        self._reset_callbacks.extend(callbacks)
+
+    def on_reset(self):
+        self._host_messages = queue.Queue()
+        for callback in self._reset_callbacks:
+            callback()
+
+    def on_hosts_updated(self, timestamp):
+        self._host_messages.put(timestamp)
+
+    def commit(self):
+        """Commits all modifications to state tracked by this object to host memory.
+
+        This call will also check for any changes to known hosts, and raise a `HostsUpdatedInterrupt`
+        if any were detected.
+
+        Because commits are a heavy operation involving data copy (potentially from GPU to host), it is
+        recommended to consider committing less frequently than once per batch. This allows users to tradeoff
+        between per-batch execution time and lost training steps in the event of a worker failure.
+        """
+        self.save()
+        self.check_host_updates()
+
+    def check_host_updates(self):
+        """Checks that a notification has been sent indicating that hosts can be added or will be removed.
+
+        Raises a `HostsUpdatedInterrupt` if such a notification has been received.
+        """
+        # Iterate through the update messages sent from the server. If the update timestamp
+        # is greater than the last update timestamp, then trigger a HostsUpdatedException.
+        last_updated_timestamp = prev_timestamp = self._last_updated_timestamp
+        while not self._host_messages.empty():
+            timestamp = self._host_messages.get()
+            if timestamp > last_updated_timestamp:
+                last_updated_timestamp = timestamp
+
+        # In order to ensure all workers raise the exception at the same time, we need to sync
+        # the updated state across all the workers.
+        # TODO(travis): this should be a max allreduce to account for changes in rank 0
+        prev_timestamp, self._last_updated_timestamp = self._bcast_object((prev_timestamp, last_updated_timestamp))
```

Comment:
I am getting a `RuntimeError` in `_bcast_object` with torch when a node in the cluster went down. This is not handled by the `@run` main loop as it only handles `HorovodInternalError` and `HostsUpdatedInterrupt`:
```
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/data/enrico/Work/git/horovod-GR-2/horovod/spark/task/gloo_exec_fn.py", line 30, in <module>
    main(codec.loads_base64(sys.argv[1]), codec.loads_base64(sys.argv[2]))
  File "/data/enrico/Work/git/horovod-GR-2/horovod/spark/task/gloo_exec_fn.py", line 23, in main
    task_exec(driver_addresses, settings, 'HOROVOD_RANK', 'HOROVOD_LOCAL_RANK')
  File "/data/enrico/Work/git/horovod-GR-2/horovod/spark/task/__init__.py", line 61, in task_exec
    result = fn(*args, **kwargs)
  File "/data/enrico/Work/git/horovod-GR-2/test/test_spark_torch.py", line 569, in fn
    return train(state, dir)
  File "/data/enrico/Work/git/horovod-GR-2/horovod/common/elastic.py", line 159, in wrapper
    return func(state, *args, **kwargs)
  File "/data/enrico/Work/git/horovod-GR-2/test/test_spark_torch.py", line 538, in train
    state.commit()
  File "/data/enrico/Work/git/horovod-GR-2/horovod/common/elastic.py", line 75, in commit
    self.check_host_updates()
  File "/data/enrico/Work/git/horovod-GR-2/horovod/common/elastic.py", line 93, in check_host_updates
    prev_timestamp, self._last_updated_timestamp = self._bcast_object((prev_timestamp, last_updated_timestamp))
  File "/data/enrico/Work/git/horovod-GR-2/horovod/torch/functions.py", line 219, in broadcast_object
    broadcast_(sz, root_rank, name + '.sz')
  File "/data/enrico/Work/git/horovod-GR-2/horovod/torch/mpi_ops.py", line 456, in broadcast_
    handle = broadcast_async_(tensor, root_rank, name)
  File "/data/enrico/Work/git/horovod-GR-2/horovod/torch/mpi_ops.py", line 434, in broadcast_async_
    return _broadcast_async(tensor, tensor, root_rank, name)
  File "/data/enrico/Work/git/horovod-GR-2/horovod/torch/mpi_ops.py", line 344, in _broadcast_async
    tensor, output, root_rank, name.encode() if name is not None else _NULL)
RuntimeError: Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
```

I am happy with getting this fixed in a separate PR after merging this beast into master.

Reply:
Nice catch, added some additional checks to raise `HorovodInternalError` when we call into the Horovod lib, which should take care of this.